nengo_dl/tensor_node.py

```diff
@@ -264,113 +362,118 @@ def build_step(self, signals):
             dtype=signals.dtype,
         )
 
-        # move minibatch dimension back to end
-        output = tf.transpose(a=output, perm=(1, 0))
-
         signals.scatter(self.dst_data, output)
 
-    def build_post(self, ops, signals, sess, rng):
-        if hasattr(self.func, "post_build"):
-            self.func.post_build(sess, rng)
 
+class Layer:
+    """
+    A wrapper for constructing TensorNodes.
+
+    This is designed to mimic and integrate with the ``tf.keras.layers.Layer`` API, e.g.
 
-def reshaped(shape_in):
-    """A decorator to reshape the inputs to a function into non-vector shapes.
+    .. testcode::
 
-    The output of the function will be flatten back into (batched) vectors.
+        with nengo.Network():
+            a = nengo.Ensemble(10, 1)
+            b = nengo_dl.Layer(tf.keras.layers.Dense(units=10))(a)
+            c = nengo_dl.Layer(lambda x: x + 1)(b)
+            d = nengo_dl.Layer(nengo.LIF())(c)
 
     Parameters
     ----------
-    shape_in : tuple of int
-        The desired shape for inputs to the function (not including the first
-        dimension, which corresponds to the batch axis)
-
-    Returns
-    -------
-    reshaper : callable
-        The decorated function
+    layer_func : callable or ``tf.keras.Layer`` or `~nengo.neurons.NeuronType`
+        A function or Keras Layer that takes the value from an input (represented
+        as a ``tf.Tensor``) and maps it to some output value, or a Nengo neuron type
+        (which will be instantiated in a Nengo Ensemble and applied to the input).
     """
 
-    def reshape_dec(func):
-        def reshaped_func(t, x):
-            batch_size = x.get_shape()[0]
-            x = tf.reshape(x, (batch_size,) + shape_in)
-            x = func(t, x)
-            x = tf.reshape(x, (batch_size, -1))
-            return x
-
-        return reshaped_func
-
-    return reshape_dec
-
-
-def tensor_layer(
-    input,
-    layer_func,
-    shape_in=None,
-    synapse=None,
-    transform=1,
-    return_conn=False,
-    # TODO: remove once there is a black release with this
-    #  bugfix https://github.com/psf/black/pull/763
-    # fmt: off
-    **layer_args
-    # fmt: on
-):
-    """A utility function to construct TensorNodes that apply some function
-    to their input (analogous to the ``tf.layers`` syntax).
+    def __init__(self, layer_func):
+        self.layer_func = layer_func
+
+    def __call__(
+        self,
+        input,
+        transform=1,
+        shape_in=None,
+        synapse=None,
+        return_conn=False,
+        # TODO: remove once there is a black release with this
```

Comment:
There's a new black release, so this might be fixed now?

Reply:
Looks like it is 👍